### PR TITLE
Move parsing-related dependencies to peer dependencies.

### PR DESCRIPTION
**Pull request by kylemh**
https://github.com/prettier/prettier-eslint/issues/402

### CODE REVIEW

**Comment by zimme** [comment]
All the "extra" parsers should also be marked as optional peer dependencies using `peerDependenciesMeta`.

File: README.md [hunk 1]
```diff
@@ -319,39 +338,41 @@ Contributions of any kind welcome!
 
 MIT
 
-[prettier]: https://github.com/jlongster/prettier
-[eslint]: http://eslint.org/
-[fix]: http://eslint.org/docs/user-guide/command-line-interface#fix
-[npm]: https://www.npmjs.com/
-[node]: https://nodejs.org
+[all-contributors]: https://github.com/kentcdodds/all-contributors
```
Comment (hamzahamidi) [inline README.md:341]: Not sure why you changed the order of the links

**Review by hamzahamidi — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (kylemh) [reply README.md:341]: Alphabetized. No real reason for it. Just felt awkward slapping the first-defined links at the end of the list.

**Comment by zimme** [comment]
Just a quick note on dependencies in npm and my view of it.

For this package I don't really think peer dependencies is the way to go because of my understanding of peer deps.

If someone don't have prettier or eslint in their project, I want them to just be able to install this project and get the latest supported versions of eslint and prettier automatically.

With npm 7+ this will actually be the case when using peer deps as they will be installed automatically unless they're marked as optional.

However, my reservation of using peer deps is also based on the fact that this package doesn't produce any exports that are intended to be used by prettier or eslint unlike eslint-config-prettier.
This package actually depends on prettier and eslint and needs them to be installed and if they're not in the project should install them. We should however maybe be less restrictive with the version that we depend on if we can't be bothered to update it in a timely manner.

So from my understanding peer deps isn't about "bring your own <package>" but more about "I produce something that you will use in <package> so you need to use it in your project". It's a small difference but still a difference.

Then a note on the optional dependencies. I don't really see what the problem with using optional dependencies based on your comment in #402.

**Comment by kylemh** [comment]
> If someone don't have prettier or eslint in their project, I want them to just be able to install this project and get the latest supported versions of eslint and prettier automatically.

I feel like this is problematic, because we'll be left - as maintainers - to constantly update the shipped dependencies.

> With npm 7+ this will actually be the case when using peer deps as they will be installed automatically unless they're marked as optional.

With every other package manager, this is NOT the case. It was actually a large point of contention: https://twitter.com/housecor/status/1325524945334661121
https://twitter.com/arcanis/status/1325727984238665728

> Then a note on the optional dependencies. I don't really see what the problem with using optional dependencies based on your comment in #402.

Aside from my comment in #402, semantically "optionalDependencies" is incorrect.

> If a dependency can be used, but you would like npm to proceed if it cannot be found or fails to install, then you may put it in the `optionalDependencies` object.

Our project cannot work without those dependencies.

**Comment by zimme** [comment]
This project can work without the ts and vue parser and those were the ones I was suggesting to go in optionalDependencies.

**Comment by kylemh** [comment]
Oh! Doi. Yes, I agree with that. I'll change it.

**Comment by kylemh** [comment]
Oh! I didn't realize. Okay, I agree those are optional, but I feel like we still have it right as-is.

Let's say somebody wants to use their own version of `vue-eslint-parser`. If we moved it to optional, wouldn't they be required to do `npm i --no-optional` which makes decisions about other dependencies for them?

**Comment by zimme** [comment]
Nope, it's an optional dependency for this package so this package won't crash and burn if it's not present, if they add it as a (dev)dependency in their project which is using this package, then we'll be able to pick it up and use it too.

However I do believe you might need to look over the require logic for those dependencies if you're changing them to be optional (peer) dependencies as the current logic assumes it will be able to import them?

**Comment by kylemh** [comment]
```js
  if (['.ts', '.tsx'].includes(fileExtension)) {
    formattingOptions.eslint.parser =
      formattingOptions.eslint.parser ||
      require.resolve('@typescript-eslint/parser');
  }

  if (['.vue'].includes(fileExtension)) {
    formattingOptions.eslint.parser =
      formattingOptions.eslint.parser || require.resolve('vue-eslint-parser');
  }
```

I've got nothing on this 😬 

You're suggesting this might not work if I move them to optional deps?

**Comment by zimme** [comment]
It will try and use any parser that the user provices via options and fallback to using the one we've installed for them, as they're currently not optional, the require.resolve call will throw an error.

>If the module can not be found, a MODULE_NOT_FOUND error is thrown.
From the require.resolve docs.

So this would need to be surrounded by a try catch call that just ignores the MODULE_NOT_FOUND error.

**Comment by zimme** [comment]
Pretty sure the commit that wraps require.resolve with try catch also removes the ability for users to specify their own parser.

**Comment by kylemh** [comment]
Shit. Duh. Sorry. Fix in the morning.

**Comment by kylemh** [comment]
Some updated dependency is logging a warning in test:

`No parser and no filepath given, using 'babel' the parser now but this will throw an error in the future. Please specify a parser or a filepath so one can be inferred.`

**Comment by zimme** [comment]
I believe we have a test that does not provide either a filepath or a parser, so prettier logs this warning as it can't use a specific parser or infer the parser based on file extension.

Not sure if this is a prettier 1 or 2 warning though and maybe we should remove support for processing without either parser or filepath.

**Comment by zimme** [comment]
I think this was supported to allow a new file in an editor, which haven't been saved yet and no syntax selected to be formatted with prettier as if it were a js file

**Comment by kylemh** [comment]
There's also still something wrong with this PR 🤔 

On `master` `npm run test` doesn't fail, but seems to have the same output. Will dig in more.

**Comment by kylemh** [comment]
It's because of missing coverage, and I'm not quite sure how to get those last bits of coverage. Comments in test code. I'll take another stab at this and the breaking change some time later!

**Pull request closed without merging**